File: windows-app/Plantoir.Tests/ClassFolderContractTests.cs
using System.Text.Json.Nodes;
using Plantoir.Core.Models;

namespace Plantoir.Tests;

/// <summary>
/// The class-folder rule, run against the SHARED contract.
///
/// <para>The cases are deserialised from <c>contracts/class-planning.json</c> →
/// <c>classFolder</c>, never retyped — the same data the macOS suite and
/// <c>scripts/test_class_folder.py</c> run against their own implementations.
/// That is the point: this rule used to exist four times and disagree, and this
/// app's copy was the one that tested the whole directory string, so a teacher
/// whose working folder was <c>C:\Users\x\Classroom\</c> made every page in
/// every course a class page.</para>
/// </summary>
public class ClassFolderContractTests
{
    [Fact]
    public void ClassFolderNaming_MatchesContract()
    {
        var doc = ContractLoader.LoadJson("class-planning.json");
        var cases = doc["classFolder"]!["naming"]!["cases"]!.AsArray();
        Assert.True(cases.Count >= 5, "the contract lost naming cases");

        foreach (var c in cases)
        {
            if (c is null) continue;
            var folders = c["perSectionFolders"]!.AsArray().Select(x => x!.ToString()).ToList();
            string expected = c["expect"]!.ToString();
            string name = c["name"]?.ToString() ?? "unnamed";

            Assert.True(ClassFolderRule.Name(folders) == expected,
                $"{name}: expected '{expected}', got '{ClassFolderRule.Name(folders)}'");
        }
    }

    [Fact]
    public void IsClassPage_MatchesContract()
    {
        var doc = ContractLoader.LoadJson("class-planning.json");
        var cases = doc["classFolder"]!["isClassPage"]!["cases"]!.AsArray();
        Assert.True(cases.Count >= 9, "the contract lost isClassPage cases");

        foreach (var c in cases)
        {
            if (c is null) continue;
            string path = c["path"]!.ToString();
            string folder = c["classFolder"]!.ToString();
            bool expected = c["expect"]!.GetValue<bool>();
            string name = c["name"]?.ToString() ?? "unnamed";

            Assert.True(ClassFolderRule.IsClassPage(path, folder) == expected,
                $"{name}: {path} against '{folder}' should be {expected}");
        }
    }

    /// <summary>
    /// The two bugs the rule was written to close, asserted directly as well as
    /// through the case list — so deleting a contract case cannot quietly
    /// delete the protection with it.
    /// </summary>
    [Fact]
    public void AShippedPageNamedForAClassIsNotALesson()
    {
        foreach (var page in new[]
                 {
                     "Setup/How This Class Works.md",
                     "Setup/Our Classroom Norms.md",
                     "Curriculum/B3. Connections Beyond the Classroom.md",
                 })
        {
            Assert.False(ClassFolderRule.IsClassPage(page, "All Classes"), page);
        }
    }

    [Fact]
    public void AWorkingFolderNamedClassroomCannotMakeEveryPageALesson()
    {
        // The path handed to the rule is RELATIVE, which is what makes the
        // teacher's own filing unable to reach it. This is the regression that
        // matters most on this platform.
        Assert.False(ClassFolderRule.IsClassPage(@"Concepts\Loops.md", "All Classes"));
        Assert.True(ClassFolderRule.IsClassPage(@"All Classes\Unit 1, Day 1.md", "All Classes"));
    }
}

File: mac-app/Tests/QuartzTeachersTests/ClassFolderContractTests.swift
import XCTest
@testable import QuartzTeachers

/// The class-folder rule, run against the SHARED contract.
///
/// The cases are deserialised from `contracts/class-planning.json` →
/// `classFolder`, never retyped — the same data `scripts/test_class_folder.py`
/// and the Windows suite run against their own implementations. That is the
/// whole point: this rule used to exist four times and disagree, and the
/// build's disagreement silently changed the Curriculum Coverage map from
/// "pages the course teaches" to "every published page".
@MainActor
final class ClassFolderContractTests: XCTestCase {

    // MARK: - Stored properties

    private var contract: [String: Any] = [:]

    // MARK: - Functions

    override func setUpWithError() throws {
        // Located the way ClassPlanningContractTests locates the same file:
        // up from #filePath to the repository root.
        let url: URL = URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent().deletingLastPathComponent()
            .deletingLastPathComponent().deletingLastPathComponent()
            .appendingPathComponent("contracts/class-planning.json")
        let all: [String: Any] = try XCTUnwrap(
            try JSONSerialization.jsonObject(with: try Data(contentsOf: url)) as? [String: Any]
        )
        contract = try XCTUnwrap(all["classFolder"] as? [String: Any],
                                 "contracts/class-planning.json has no classFolder section")
    }

    func testNamingCasesFromTheContract() throws {
        let naming: [String: Any] = try XCTUnwrap(contract["naming"] as? [String: Any])
        let cases: [[String: Any]] = try XCTUnwrap(naming["cases"] as? [[String: Any]])
        XCTAssertGreaterThanOrEqual(cases.count, 5, "the contract lost naming cases")
        for testCase in cases {
            let name: String = (testCase["name"] as? String) ?? "unnamed"
            let folders: [String] = (testCase["perSectionFolders"] as? [String]) ?? []
            let expected: String = try XCTUnwrap(testCase["expect"] as? String)
            XCTAssertEqual(
                ClassFolder.name(inPerSectionFolders: folders), expected,
                "\(name): \((testCase["why"] as? String) ?? "")"
            )
        }
    }

    func testIsClassPageCasesFromTheContract() throws {
        let rule: [String: Any] = try XCTUnwrap(contract["isClassPage"] as? [String: Any])
        let cases: [[String: Any]] = try XCTUnwrap(rule["cases"] as? [[String: Any]])
        XCTAssertGreaterThanOrEqual(cases.count, 9, "the contract lost isClassPage cases")
        for testCase in cases {
            let name: String = (testCase["name"] as? String) ?? "unnamed"
            let path: String = try XCTUnwrap(testCase["path"] as? String)
            let folder: String = try XCTUnwrap(testCase["classFolder"] as? String)
            let expected: Bool = try XCTUnwrap(testCase["expect"] as? Bool)
            XCTAssertEqual(
                ClassFolder.isClassPage(relativePath: path, classFolder: folder), expected,
                "\(name): \((testCase["why"] as? String) ?? "")"
            )
        }
    }

    /// The two bugs the rule was written to close, asserted directly as well as
    /// through the case list — so deleting a contract case cannot quietly
    /// delete the protection with it.
    func testAShippedPageNamedForAClassIsNotALesson() {
        let pages: [String] = [
            "Setup/How This Class Works.md",
            "Setup/Our Classroom Norms.md",
            "Curriculum/B3. Connections Beyond the Classroom.md",
        ]
        for page in pages {
            XCTAssertFalse(
                ClassFolder.isClassPage(relativePath: page, classFolder: "All Classes"),
                "\(page) is not a lesson — it ships in the example content"
            )
        }
    }

    func testAWindowsPathSeparatorIsUnderstood() {
        XCTAssertTrue(ClassFolder.isClassPage(
            relativePath: "All Classes\\Unit 1, Day 1.md", classFolder: "All Classes"))
    }
}
